### PR TITLE
[Enhancement] support array_generate for date type.

### LINKED ISSUE #55099: Support date type as input parameter of array_generate like trino sequence
## Feature request

**Is your feature request related to a problem? Please describe.**

trino sequence function
select sequence(current_date - interval '2' day, current_date) as a -- return ['2025-01-13', '2025-01-14', '2025-01-15']

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by zhangheihei**
## Why I'm doing:
Fixes #55099
## What I'm doing:

select array_generate("2025-10-01", "2027-10-05",interval 1 year);
select array_generate("2025-10-01", "2027-10-05",interval 1 QUARTER);

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.0
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

---

> [!NOTE]
> Adds DATE/DATETIME `array_generate` with interval units (year→microsecond), including FE parsing/validation and BE chrono-based execution, plus tests.
> 
> - **Backend**:
>   - Add interval-based generators for `array_generate` with units: `year|quarter|month|week|day|hour|minute|second|millisecond|microsecond`.
>   - Introduce `calculate_accurate_step_count` and per-unit dispatch functions; use C++20 chrono for precise time arithmetic.
>   - Add `ArrayGenerate::prepare/close` for DATE/DATETIME to choose unit-specific impl at runtime.
>   - Extend `DateValue` and `TimestampValue` with `to_sys_days()` and `to_sys_time()` helpers.
> - **Frontend**:
>   - Parser: support `array_generate(start, stop, INTERVAL n UNIT)` and normalize step to `INT` plus unit `VARCHAR`.
>   - Analyzer: detect date/datetime signatures, enforce constant non-negative step, default unit to `day`, and switch function resolution to non-strict supertype.
> - **Function registry**:
>   - Register DATE/DATETIME variants of `array_generate` with prepare/close hooks in `gensrc/script/functions.py`.
> - **Tests**:
>   - Add BE unit tests for DATE/DATETIME across all units, directions, nulls, and edge cases.
>   - Add SQL tests validating outputs, errors, and interval syntax.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 1349b7adbb5d2275f1cdef265178a1541af07d15. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by alvin-phoenix-ai** [comment]
@cursor review

File: be/src/exprs/array_functions.tpp [hunk 1]
```diff
@@ -1644,114 +1644,414 @@ private:
...
-        auto dst = ArrayColumn::create(std::move(array_elements), std::move(array_offsets));
+            auto dst = ArrayColumn::create(std::move(array_elements), std::move(array_offsets));
 
-        if (all_const_cols) {
-            if (nulls->is_null(0)) {
-                return ColumnHelper::create_const_null_column(num_rows);
-            } else {
-                return ConstColumn::create(std::move(dst), num_rows);
+            if (all_const_cols) {
+                if (nulls->is_null(0)) {
```
[the remark this answers is not in the document]
Comment (zhangheihei) [reply be/src/exprs/array_functions.tpp:2075]: nuls must not be nullptr

**Comment by alvin-phoenix-ai** [comment]
@cursor review

File: be/src/exprs/array_functions.tpp [hunk 2]
```diff
@@ -1644,114 +1644,414 @@ private:
     }
 };
 
+// Define template function for array generation with different time units
+// Constants for time unit estimation
+constexpr int64_t DAYS_PER_YEAR = 365;
+constexpr int64_t DAYS_PER_QUARTER = 90;
```
Comment (trueeyu) [inline be/src/exprs/array_functions.tpp]: Is it inaccurate to use something like DAYS_PER_QUARTER? It’s better to implement it with a time library, right?

**Comment by alvin-phoenix-ai** [comment]
@cursor review

**Comment by alvin-phoenix-ai** [comment]
@cursor review

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java [hunk 3]
```diff
@@ -1067,6 +1067,9 @@ public Void visitFunctionCall(FunctionCallExpr node, Scope scope) {
             Type[] argumentTypes = node.getChildren().stream().map(Expr::getType).toArray(Type[]::new);
             // check fn & throw exception direct if analyze failed
             checkFunction(fnName, node, argumentTypes);
+            // checkFunction may modify node's children (e.g., add default parameters for array_generate),
```
Comment (trueeyu) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java]: Why update argumentTypes at checkFunction?

File: be/src/exprs/array_functions.tpp [hunk 4]
```diff
@@ -1644,114 +1645,446 @@ private:
...
+        } else if constexpr (TIME_UNIT == TimeUnit::QUARTER) {
+            // Quarter difference: years * 4 + quarter_offset
+            int years_diff = static_cast<int>(stop_ymd.year()) - static_cast<int>(start_ymd.year());
+            int start_quarter = (static_cast<unsigned>(start_ymd.month()) - 1) / 3;
+            int stop_quarter = (static_cast<unsigned>(stop_ymd.month()) - 1) / 3;
+            diff = std::abs(years_diff * 4 + (stop_quarter - start_quarter));
+        } else { // MONTH
+            // Month difference: years * 12 + month_offset
+            int years_diff = static_cast<int>(stop_ymd.year()) - static_cast<int>(start_ymd.year());
+            int months_diff = static_cast<unsigned>(stop_ymd.month()) - static_cast<unsigned>(start_ymd.month());
```
[the remark this answers is not in the document]
Comment (zhangheihei) [reply be/src/exprs/array_functions.tpp:1738]: There are 12 months in total; no overflow occurs.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java (hunk 3 above)
Comment (zhangheihei) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java]: The logic for updating argumentTypes has been moved to FunctionAnalyzer.getAnalyzedFunction.

**Comment by andyziye** [comment]
@Mergify rebase

**Review by Seaven — DISMISSED** [review]

File: be/src/exprs/array_functions.tpp [hunk 5]
```diff
@@ -1644,114 +1645,446 @@ private:
...
+    }
+
+    return 0;
+}
+
+#define DEFINE_ARRAY_GENERATE_FN(NAME, TIME_UNIT)                                                                  \
+    template <LogicalType LType, LogicalType ResultType>                                                           \
+    static StatusOr<ColumnPtr> array_generate_function_##NAME(FunctionContext* ctx, const Columns& columns) {      \
+        RETURN_IF_COLUMNS_ONLY_NULL(columns);                                                                      \
+        DCHECK(columns.size() == 4);                                                                               \
```
Comment (silverbullet233) [inline be/src/exprs/array_functions.tpp]: it's better to return an error if columns' size is mismatched so that we can avoid some undefined behavior

**Comment by alvin-phoenix-ai** [comment]
@cursor review

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java [hunk 6]
```diff
@@ -7704,6 +7704,23 @@ public ParseNode visitSimpleFunctionCall(com.starrocks.sql.parser.StarRocksParse
 
         FunctionName fnName = FunctionName.createFnName(fullFunctionName);
         String functionName = fnName.getFunction();
+        if (functionName.equals(FunctionSet.ARRAY_GENERATE)) {
```
Comment (trueeyu) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:7707]: What will happen if step is 0 or < 0?

**Review by trueeyu — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]
